File: android/src/main/java/com/example/ivs_broadcaster/streams/StageController.java
package com.example.ivs_broadcaster.streams;

import android.annotation.SuppressLint;
import android.content.Context;
import android.os.Build;
import android.util.Log;

import androidx.annotation.NonNull;
import androidx.annotation.RequiresApi;

import com.amazonaws.ivs.broadcast.Stage;
import com.example.ivs_broadcaster.chats.StageChat;
import com.example.ivs_broadcaster.streams.strategy.PublishStrategy;
import com.example.ivs_broadcaster.streams.strategy.ViewerStrategy;

import io.flutter.embedding.engine.plugins.FlutterPlugin;
import io.flutter.plugin.common.EventChannel;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;

@SuppressLint("NewApi")
public class StageController {
    Stage stage;
    FlutterPlugin.FlutterPluginBinding binding;
    EventChannel.EventSink stageSink;
    StageChat stageChat;

    public StageController(FlutterPlugin.FlutterPluginBinding binding, EventChannel.EventSink stageSink, StageChat stageChat) {
        this.binding = binding;
        this.stageSink = stageSink;
        this.stageChat = stageChat;
    }

    public void joinStage(MethodCall methodCall, @NonNull MethodChannel.Result result) {
        String token = (String) methodCall.argument("token");
        boolean shouldPublish = (boolean) methodCall.argument("shouldPublish");
        Log.d("StageController", token);
        Log.d("StageController", String.valueOf(shouldPublish));
        if (shouldPublish) {
            stage = new Stage(binding.getApplicationContext(), token, new PublishStrategy(binding.getApplicationContext()));
        } else {
            stage = new Stage(binding.getApplicationContext(), token, new ViewerStrategy());
        }
        stage.addRenderer(new StageListener(binding, stageSink, stageChat));
        try {
            stage.join();
            Log.d("StageController", "Stage is joined");
        } catch (Exception e) {
            Log.d("StageController", e.getMessage());
        }
        result.success("Success");
    }

    public void leaveStage(MethodCall methodCall, @NonNull MethodChannel.Result result) {
        stage.leave();
        stage.release();
        result.success("StageController is leaving...");
    }
}

File: ios/Classes/streams/StageController.swift
import Flutter
import AmazonIVSBroadcast

public class StageController: NSObject {
    public var stage: IVSStage?;
    public var stageSink: FlutterEventSink?
    private var stageChat: StageChat?
    private var binding: FlutterPluginRegistrar
    private var viewer: ViewerStrategy
    private var publiser: PublishStrategy
    private var listener: StageListener

    init(stageSink: FlutterEventSink? = nil, stageChat: StageChat? = nil, binding:FlutterPluginRegistrar) {
        self.binding = binding
        self.stageSink = stageSink
        self.stageChat = stageChat
        self.viewer = ViewerStrategy()
        self.publiser = PublishStrategy()
        self.listener = StageListener(stageSink: stageSink, stageChat: stageChat!, binding: binding)
    }
    
    public func joinStage(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        if let arguments = call.arguments as? [String: Any] {
            let token = arguments["token"] as? String
            let shouldPublish = arguments["shouldPublish"] as? Bool ?? false
            
            print("StageController: token \(token!)")
            print("StageController: shouldPublish \(shouldPublish)")
            
            do {
                if shouldPublish {
                    stage = try IVSStage(token: token!, strategy: publiser);
                    print("StageController: published")
                }else {
                    stage = try IVSStage(token: token!, strategy: viewer);
                    print("StageController: viewer")
                }
                do {
                    stage?.leave()
                    stage!.addRenderer(listener)
                    let x = try stage?.join()
                    print("StageController: stage is joined \(x)")
                } catch {
                    print("StageController: stage is not joined")
                }
                print("StageController: Set stage")
            } catch {
                print("StageController: Can't connect to stage")
            }
        }
        result("Success")
    }
    
    public func leaveStage(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        stage?.leave()
        print("StageController: Stage is leaving")
        result("StageController: StageController is leaving...")
    }
}

